crates/accelerate/src/circuit_library/iqp.rs

```diff
@@ -0,0 +1,159 @@
+    // triangular matrix. This contains N(N+1)/2 unique random values, stored as
+    //    [v_0 v_1 ... v_{N-1}]
+    //    [    v_N ... v_{2N-2}]
+    //             ...
+    //    [            v_{-1}]
+    let num_qubits = num_qubits as usize;
+    let num_values = num_qubits * (num_qubits + 1) / 2;
+    let values: Vec<i64> = (0..num_values)
+        .map(|_| rand::thread_rng().gen_range(0..8) as i64)
+        .collect();
+
+    // We then build the matrix of dimension NxN by reading from the ``values`` vector.
+    // * since the matrix is symmetric, we treat (i, j) and (j, i) equally, which we implement
+    //   here by sorting the indices before accessing the value vector
+    // * in each row, the offset with which we access ``values`` changes: in row K, we need to
+    //   start reading from index: \sum_{k=0}^K N-k
+    Array2::from_shape_fn((num_qubits, num_qubits), |(i, j)| {
+        let (low, high) = if i < j { (i, j) } else { (j, i) };
+        let offset = low * (2 * num_qubits - low + 1) / 2;
+        values[offset + high - low]
+    })
+}
+
+/// Returns true if the input matrix is symmetric, otherwise false.
+fn check_symmetric(matrix: &ArrayView2<i64>) -> bool {
+    let nrows = matrix.nrows();
+
+    if matrix.ncols() != nrows {
+        return false;
+    }
+
+    for i in 0..nrows {
+        for j in i + 1..nrows {
+            if matrix[(i, j)] != matrix[(j, i)] {
+                return false;
+            }
+        }
+    }
+
+    true
+}
+
+/// Implement an Instantaneous Quantum Polynomial time (IQP) circuit.
+///
+/// This class of circuits is conjectured to be classically hard to simulate,
+/// forming a generalization of the Boson sampling problem. See Ref. [1] for
+/// more details.
+///
+/// Args:
+///     num_qubits: The number of qubits in the IQP circuit.
+///     interactions: If provided, this is a symmetric square matrix of width ``num_qubits``,
+///         determining the operations in the IQP circuit. The diagonal represents the power
+///         of single-qubit T gates and the upper triangular part the power of CS gates
+///         in between qubit pairs. If None, a random interactions matrix will be sampled.
+///
+/// References:
+///
+///     [1] M. J. Bremner et al. Average-case complexity versus approximate simulation of
+///     commuting quantum computations, Phys. Rev. Lett. 117, 080501 (2016).
+///     `arXiv:1504.07999 <https://arxiv.org/abs/1504.07999>`_
+#[pyfunction]
+#[pyo3(signature = (num_qubits, interactions=None))]
+pub fn py_iqp(
+    py: Python,
+    num_qubits: u32,
+    interactions: Option<PyReadonlyArray2<i64>>,
+) -> PyResult<CircuitData> {
+    let array = match interactions {
+        Some(matrix) => {
+            let view = matrix.as_array();
+            if !check_symmetric(&view) {
+                return Err(CircuitError::new_err("IQP matrix must be symmetric."));
+            }
+            view.to_owned()
```

Comment:
Do we need to copy the interaction matrix here?

Reply:
Hmm no we shouldn't need to. I think I did that because Rust was unhappy about the match-arms returning different types (as `generate_random_interactions` returns an owned instance). But I should be able to refactor this by moving the function call into the arms 👍🏻

Reply:
I tried but I couldn't get it to work without cloning... I would guess that's because we create an iterator in `iqp` that needs to own the matrix to work 🙂

Reply:
Resolved with e7258c4